### PR TITLE
chore: sign container images and checksum assets

### LINKED ISSUE #2330: Create Signed Docker Images
# Summary

We should sign our docker images like argo cd has just done via this PR https://github.com/argoproj/argo-cd/pull/10925

# Use Cases

For security when end users want to know that the image has been built by our github actions.

---

**Message from the maintainers**:

Impacted by this bug? Give it a 👍. We prioritize the issues with the most 👍.

### PR DESCRIPTION
**Pull request by 34fathombelow**
Signed-off-by: Justin Marquis <34fathombelow@protonmail.com>
Closes #2330 

This PR implements container images and ```argo-rollouts-checksums.txt``` to be signed using ```sigstore/cosign```  

#### Three GitHub secrets will need to be created before this PR is merged. The process to do this is listed below.

TLDR: https://docs.sigstore.dev/cosign/git_support

1. [Install Cosign](https://docs.sigstore.dev/cosign/installation) on your workstation. Linux, Homebrew, and container options are available.  Execute the command ```cosign --version``` to verified it has been installed correctly.
2. Create or have a valid GitHub PAT token available.
3. Export your PAT as the environment variable "GITHUB_TOKEN"
4. Execute ```cosign generate-key-pair github://argoproj/argo-rollouts``` This will start the process of creating the GitHub Secrets automatically for you, and prompt you to enter a password.  This Password will be stored as the Github secret ```COSIGN_PASSWORD``` I would recommend to use well respected password generator such as KeePassXC or Bitwarden and use a paranoid level of characters of at least 32. 
5. Three GitHub secretes should have been created, please verify.  ```COSIGN_PASSWORD``` ```COSIGN_PRIVATE_KEY``` & ```COSIGN_PUBLIC_KEY```
6. Please comment on this PR with the newly created public key (cosign.pub)

### CODE REVIEW

**Review by zachaller — APPROVED** [review]

**Comment by jessesuen** [comment]
The COSIGN_PASSWORD COSIGN_PRIVATE_KEY & COSIGN_PUBLIC_KEY secrets have been added. Here is the public key:

```
-----BEGIN PUBLIC KEY-----
MFkwEwYHKoZIzj0CAQYIKoZIzj0DAQcDQgAElm4DvuUfW6x0jB8vZ5wO+XeyaAbK
3/VjNMOsZ2svjls3mc/1eg2cP3MR9kOA7HD75nERY+U82vqhGsZ8bwnMLQ==
-----END PUBLIC KEY-----
```

I've announced the above public key, but in the future this might change (if we for example switch from github actions), right?

I've discarded the private key and password so they only exist in GitHub now.

**Review by jessesuen — APPROVED** [review]
I'll merge this in after conflict is resolved. Thanks!

File: .github/workflows/docker-publish.yml
Comment (34fathombelow) [inline .github/workflows/docker-publish.yml:106]: This output provides two separate tags(master and latest) , we all need to sign by digest instead.
please do not merge till changes are made.

File: .github/workflows/docker-publish.yml
Comment (34fathombelow) [inline .github/workflows/docker-publish.yml:117]: This output provides two separate tags(master and latest) , we all need to sign by digest instead.
please do not merge till changes are made.

**Review by zachaller — APPROVED** [review]

**Comment by zachaller** [comment]
part of https://github.com/argoproj/argo-rollouts/issues/2329

**Pull request merged**
